pkg/host-disk/host-disk.go

```diff
@@ -85,19 +90,67 @@ func getPVCDiskImgPath(volumeName string) string {
 	return path.Join(pvcBaseDir, volumeName, "disk.img")
 }
 
-func CreateHostDisks(vmi *v1.VirtualMachineInstance) error {
+type DiskImgCreator struct {
+	dirBytesAvailableFunc func(path string) (uint64, error)
+	notifier              k8sNotifier
+}
+
+type k8sNotifier interface {
+	SendK8sEvent(args *notifier.K8sEventArgs) error
+}
+
+func NewHostDiskCreator(notifier k8sNotifier) DiskImgCreator {
+	return DiskImgCreator{
+		dirBytesAvailableFunc: dirBytesAvailable,
+		notifier:              notifier,
+	}
+}
+
+func (hdc DiskImgCreator) Create(vmi *v1.VirtualMachineInstance) error {
 	for _, volume := range vmi.Spec.Volumes {
 		if hostDisk := volume.VolumeSource.HostDisk; hostDisk != nil && hostDisk.Type == v1.HostDiskExistsOrCreate && hostDisk.Path != "" {
 			if _, err := os.Stat(hostDisk.Path); os.IsNotExist(err) {
-				availableSpace, err := dirBytesAvailable(path.Dir(hostDisk.Path))
+				availableSize, err := hdc.dirBytesAvailableFunc(path.Dir(hostDisk.Path))
 				if err != nil {
 					return err
 				}
-				size, _ := hostDisk.Capacity.AsInt64()
-				if uint64(size) > availableSpace {
-					return fmt.Errorf("Unable to create %s, not enough space, demanded size %d B is bigger than available space %d B", hostDisk.Path, uint64(size), availableSpace)
+				requestedSize, _ := hostDisk.Capacity.AsInt64()
+				diskSize := requestedSize
+				if uint64(requestedSize) > availableSize {
+					// Some storage provisioners provision less space than requested, due to filesystem overhead etc.
+					// We tolerate some difference in requested and available capacity up to some degree.
+					// This can be configured with the "pvc-tolerate-less-space-up-to-percent" parameter in the kubevirt-config ConfigMap.
+					// It is provided as environment variable to virt-launcher.
+					t := os.Getenv(lessPvcSpaceTolerationEnvName)
+					if t == "" {
+						// Default value is set in virt-controller, so this only happens in unit tests
+						// For toleration related tests we explicitly set a value, so let's keep this at 0 for the other tests
+						t = "0"
+					}
+					toleration, err := strconv.Atoi(t)
+					if err != nil {
+						return fmt.Errorf("Unable to create %s, invalid toleration value %v", hostDisk.Path, toleration)
+					}
+					toleratedSize := requestedSize * (100 - int64(toleration)) / 100
+					if uint64(toleratedSize) > availableSize {
+						return fmt.Errorf("Unable to create %s, not enough space, demanded size %d B is bigger than available space %d B, also after taking %d %% toleration into account",
```

Comment:
Small detail, KubeVirt project generally wants to follow the Kubernetes standard of not capitalizing error messages. This is due to the fact that they're often chained together.

Reply:
thanks, wasn't aware of this, will change it

Reply:
done